src/longrange_force_types/force_ewald_cabana_neigh_impl.h

```diff
@@ -0,0 +1,597 @@
+            {
+                x( idx, dim ) += domain_size( dim );
+            }
+        }
+
+        offset += halos.at( 2 * dim )->numGhost();
+
+        // do transfer of particles in upward direction
+        halos.at( 2 * dim + 1 ) = new Cabana::Halo<DeviceType>(
+            comm, N_max, export_ids_up, export_ranks_up, neighbors );
+        N_max += halos.at( 2 * dim + 1 )->numGhost();
+
+        // resize particle list to contain all particles
+        system->xvf.resize( N_max );
+
+        // update slices
+        x = Cabana::slice<Positions>(system->xvf);
+        f = Cabana::slice<Forces>(system->xvf);
+        //f_a = Cabana::slice<Forces>(system->xvf);
+        id = Cabana::slice<IDs>(system->xvf);
+        //type = Cabana::slice<Types>(system->xvf);
+        q = Cabana::slice<Charges>(system->xvf);
+        p = Cabana::slice<Potentials>(system->xvf);
+
+        // gather data for halo regions
+
+        Cabana::gather( *( halos.at( 2 * dim + 1 ) ), x );
+        Cabana::gather( *( halos.at( 2 * dim + 1 ) ), q );
+        Cabana::gather( *( halos.at( 2 * dim + 1 ) ), p );
+        Cabana::gather( *( halos.at( 2 * dim + 1 ) ), f );
+        Cabana::gather( *( halos.at( 2 * dim + 1 ) ), id );
+
+        // do periodic corrections and reset partial forces
+        // and potentials of ghost particles
+        // (they are accumulated during the scatter step)
+        for ( int idx = N_local + offset;
+              (std::size_t)idx <
+              N_local + offset + halos.at( 2 * dim + 1 )->numGhost();
+              ++idx )
+        {
+            p( idx ) = 0.0;
+            f( idx, 0 ) = 0.0;
+            f( idx, 1 ) = 0.0;
+            f( idx, 2 ) = 0.0;
+            if ( loc_dims.at( dim ) == 0 )
+            {
+                x( idx, dim ) -= domain_size( dim );
+            }
+        }
+
+        offset += halos.at( 2 * dim + 1 )->numGhost();
+    }
+
+    // create VerletList to iterate over
+
+    double min_coords[3];
+    double max_coords[3];
+
+    for ( int dim = 0; dim < 3; ++dim )
+    {
+        max_coords[dim] = -2.0 * r_max;
+        min_coords[dim] = domain_size( dim ) + 2.0 * r_max;
+    }
+
+    for ( int idx = 0; idx < N_max; ++idx )
+    {
+        for ( int dim = 0; dim < 3; ++dim )
+        {
+            if ( x( idx, dim ) < min_coords[dim] )
+                min_coords[dim] = x( idx, dim );
+            if ( x( idx, dim ) > max_coords[dim] )
+                max_coords[dim] = x( idx, dim );
+        }
+    }
+
+    // compute cell size in a way that not too many cells are used
+    double cell_size;
+    cell_size =
+        std::max( std::min( ( grid_max[0] - grid_min[0] ) / 20.0,
+                            std::min( ( grid_max[1] - grid_min[1] ) / 20.0,
+                                      ( grid_max[2] - grid_min[2] ) / 20.0 ) ),
+                  1.0 );
+
+    ListType verlet_list( x, 0, N_local, r_max, cell_size, grid_min, grid_max );
```

Comment:
The main reason we separate the creation of the neighbor list and the force compute is that we can get away with reusing the neighbors for ~50 steps.

Reply:
That makes sense. The original example did not really consider this separation since there were no timesteps. 

I am thinking I'll move the work done in lines 322-533 here into a function in comm called something like `void Comm::update_force()`, then add calls to it in cabanamd.cpp? Once that is set up I can see where forces should or should not be reset. Your thoughts?